### PR TITLE
US119022 Page load telemetry

### PR DESCRIPTION
**Pull request by Vitalii-Misechko**
[US119022](https://rally1.rallydev.com/#/detail/userstory/408922032584?fdp=true): [Engagement] Page Load Telemetry

Each page load causes 4 API calls to Production Telemetry Service. Each event contains `EventBody.Object.Id` which is the same across these 4 calls. Telemetry records following metrics.
```JSON
[
  "first-paint",
  "first-contentful-paint",
  "d2l.page.preFetch",
  "d2l.page.timeToClient",
  "d2l.page.webComponentsReady",
  "d2l.page.domInteractive",
  "d2l.page.load",
  "d2l.page.server",
  "d2l.page.download",
  "d2l.page.domContentLoadedHandlers",
  "d2l.page.tti"
]
```
 [GRID web performance](https://docs.dev.d2l/index.php/GRID:_Web_Performance) especially pays attention to First Contentful Paint(_first-contentful-paint_) and Time to Interactive(_d2l.page.tti_).
[Examples of performance events](https://github.com/Brightspace/insights-engagement-dashboard/files/5453985/Performance.Events.From.Client.txt)  that are sent to telemetry service

### Functional Testing - N/A
* Unit tests
* API calls to telemetry endpoint returns 200 HTTP Code
### Performance Testing
* Telemetry API calls are executed asynchronously, therefore, these calls do not affect UI .
### UX and docs - N/A

- [x] Simple queries to the telemetry data

FYI @rohitvinnakota @MykolaGalian @johngwilkinson @NicholasHallman

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -6,6 +6,27 @@
...
+
+			fetch.mock(
+				async(url, opts) => {
+					console.log(`${opts.method} ${url}`);
+					console.dir('Body: ');
+					const body = JSON.parse(await opts.body);
+					console.dir(body);
+
+					// verifies that only telemetry is sent on demo page
+					const endpoint = 'https://endpoint.com/api/events/';
```
Comment (bemailloux) [inline demo/index.html:22]: This should be `example.com`. `endpoint.com` is an actual website, so if someone deleted fetchMock by accident we'd be sending actual data to endpoint.com, which would not be good.

File: engagement-dashboard.js [hunk 2]
```diff
@@ -265,5 +282,50 @@ class EngagementDashboard extends Localizer(MobxLitElement) {
...
+			measures: [event.detail.value],
+			action: 'PageLoad'
+		});
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+
+		this._boundHandlePageLoad = this._handlePageLoad.bind(this);
+		window.addEventListener('load', this._boundHandlePageLoad);
```
Comment (bemailloux) [inline engagement-dashboard.js:318]: TIL: I was wondering when this function would run, if you added a `load` event listener after the page was loaded. Turns out `DOMContentLoaded` and `load` are different

File: model/telemetry-helper.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+	_createEvent(eventType, eventBody) {
+		return new d2lTelemetryBrowserClient.TelemetryEvent()
+			.setDate(new Date())
+			.setType(eventType)
+			.setSourceId('insights')
+			.setBody(eventBody);
+	}
+
+	logPerformanceEvent({ id, measures, action }) {
```
Comment (bemailloux) [inline model/telemetry-helper.js:18]: Should we include a list of the valid actions as a const somewhere?

File: model/telemetry-helper.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		if (!this._client || !window.location) {
+			return;
+		}
+
+		if (measures.lenght < 1) {
```
Comment (bemailloux) [inline model/telemetry-helper.js:23]: Typo here: this will always be `false`

File: model/telemetry-helper.js (hunk 3 above)
Comment (Vitalii-Misechko) [reply model/telemetry-helper.js:18]: Yes, we should, I've attached an example of Athena query to the story. In general, we can come up with any action we want. The code in this PR fires only events with `PageLoad` action

File: demo/index.html (hunk 1 above)
Comment (Vitalii-Misechko) [reply demo/index.html:22]: fixed

File: engagement-dashboard.js [hunk 5]
```diff
@@ -265,5 +282,50 @@ class EngagementDashboard extends Localizer(MobxLitElement) {
...
+		}
+
+		this._telemetryHelper.logPerformanceEvent({
+			id: this.telemetryId,
+			measures: getPerformanceLoadPageMeasures(),
+			action: 'PageLoad'
+		});
+	}
+
+	_handlePerfomanceMeasure(event) {
```
Comment (anhill-D2L) [inline engagement-dashboard.js:298]: Typo: Performance

File: engagement-dashboard.js [hunk 6]
```diff
@@ -265,5 +282,50 @@ class EngagementDashboard extends Localizer(MobxLitElement) {
...
+		});
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+
+		this._boundHandlePageLoad = this._handlePageLoad.bind(this);
+		window.addEventListener('load', this._boundHandlePageLoad);
+
+		this._boundHandlePerfomanceMeasure = this._handlePerfomanceMeasure.bind(this);
```
Comment (anhill-D2L) [inline engagement-dashboard.js:320]: Typo: Performance

File: model/telemetry-helper.js (hunk 4 above)
Comment (Vitalii-Misechko) [reply model/telemetry-helper.js]: Good catch! Fixed unit test as well

File: engagement-dashboard.js (hunk 5 above)
Comment (Vitalii-Misechko) [reply engagement-dashboard.js:298]: Eagle eye! Ha ha perfomance )

File: engagement-dashboard.js [hunk 7]
```diff
@@ -265,5 +282,50 @@ class EngagementDashboard extends Localizer(MobxLitElement) {
...
+			action: 'PageLoad'
+		});
+	}
+
+	_handlePerfomanceMeasure(event) {
+		if (!this._telemetryHelper) {
+			return;
+		}
+
+		if (!['d2l.page.tti', 'first-paint', 'first-contentful-paint'].includes(event.detail.value.name)) {
```
Comment (anhill-D2L) [inline engagement-dashboard.js:303]: Just curious: Where will/are the names of the events being set?

Comment (Vitalii-Misechko) [reply engagement-dashboard.js:303]: I believe that LMS at some point injects JS code into a page that is similar to what you can [see here](https://docs.dev.d2l/index.php/GRID:_Web_Performance#PerformanceObserver). 
`d2l.page.tti` you can find [here](http://old-search.dev.d2l/source/xref/Lms/lp/framework/web/D2L.LP.Web/UI/PerformanceMeasures/Default/TimeToInteractivePolyfillPageElementLoader.cs#18)
However, I can't find others

File: engagement-dashboard.js (hunk 6 above)
Comment (Vitalii-Misechko) [reply engagement-dashboard.js]: fixed

File: engagement-dashboard.js (hunk 5 above)
Comment (Vitalii-Misechko) [reply engagement-dashboard.js]: fixed

**Comment by Vitalii-Misechko** [comment]
FYI
I've added to the story examples of queries and queries results.

**Review by bemailloux — APPROVED** [review]

**Comment by anhill-D2L** [comment]
> FYI
> I've added to the story examples of queries and queries results.

It looks like the story has names of links rather than actual links.

A doc in this repo about querying for telemetry would probably be useful and easier to find

**Review by anhill-D2L — APPROVED** [review]

**Comment by Vitalii-Misechko** [comment]
> > FYI
> > I've added to the story examples of queries and queries results.
> 
> It looks like the story has names of links rather than actual links.

The `Notes` section contains descriptions and the `Attachments` section contains mentioned queries and results

> A doc in this repo about querying for telemetry would probably be useful and easier to find

Good call! It seems that I need to gather info from 3 telemetry stories somewhere.
Should I choose confluence or docs dev instead of MD-file in this public available repo?

**Comment by anhill-D2L** [comment]
> Good call! It seems that I need to gather info from 3 telemetry stories somewhere.
> Should I choose confluence or docs dev instead of MD-file in this public available repo?

I forgot that this repo is public

I'd prefer confluence over docs.dev

**Comment by bemailloux** [comment]
> I forgot that this repo is public

Wait a second - should it be public?

**Comment by johngwilkinson** [comment]
> > > FYI
> > > I've added to the story examples of queries and queries results.
> > 
> > 
> > It looks like the story has names of links rather than actual links.
> 
> The `Notes` section contains descriptions and the `Attachments` section contains mentioned queries and results
> 
> > A doc in this repo about querying for telemetry would probably be useful and easier to find
> 
> Good call! It seems that I need to gather info from 3 telemetry stories somewhere.
> Should I choose confluence or docs dev instead of MD-file in this public available repo?

I'm not sure if there's any reason for this repo to be public - maybe we should make it private.

I certainly prefer having docs live as close to the code as possible.

**Comment by Vitalii-Misechko** [comment]
> > > > FYI
> > > > I've added to the story examples of queries and queries results.
> > > 
> > > 
> > > It looks like the story has names of links rather than actual links.
> > 
> > 
> > The `Notes` section contains descriptions and the `Attachments` section contains mentioned queries and results
> > > A doc in this repo about querying for telemetry would probably be useful and easier to find
> > 
> > 
> > Good call! It seems that I need to gather info from 3 telemetry stories somewhere.
> > Should I choose confluence or docs dev instead of MD-file in this public available repo?
> 
> I'm not sure if there's any reason for this repo to be public - maybe we should make it private.
> 
> I certainly prefer having docs live as close to the code as possible.

I like this approach as well. 

We had [the discussion public vs private earlier](https://d2l.slack.com/archives/CMNGJ22G0/p1595247657281300). Probably we should review it. I believe that we can turn this repo into private, but we need to solve integration with BSI question.

I'm going to write some doc about telemetry in confluence and make a reference from readme.md of this repo. If we decide to make this repo private I'll move telemetry documentation right into it.

**Comment by bemailloux** [comment]
> We had the discussion public vs private earlier. Probably we should review it. I believe that we can turn this repo into private, but we need to solve integration with BSI question.

Forgot about that :) that's a good reason to keep the repo public

**Comment by Vitalii-Misechko** [comment]
PR for telemetry documentation in README.md
https://github.com/Brightspace/insights-engagement-dashboard/pull/128

**Pull request merged**